File: Binary Tree/bfs.py
from collections import deque

class Node:
    def __init__(self, val):
        self.data = val
        self.left = None
        self.right = None

class Tree:
    def __init__(self):
        self.root = None

    def build_tree(self, s):
        """Builds a binary tree from a space-separated level-order string,
        where 'N' denotes a null."""
        if not s or s[0] == 'N':
            return

        vals = s.split()
        self.root = Node(int(vals[0]))
        queue = deque([self.root])
        i = 1

        while queue and i < len(vals):
            curr = queue.popleft()

            # left child
            if vals[i] != 'N':
                curr.left = Node(int(vals[i]))
                queue.append(curr.left)
            i += 1

            # right child
            if i < len(vals) and vals[i] != 'N':
                curr.right = Node(int(vals[i]))
                queue.append(curr.right)
            i += 1

    def bfs(self):
        """Prints the tree in level-order."""
        if not self.root:
            print("Tree is empty.")
            return

        queue = deque([self.root])
        print("Level-order Traversal:", end=" ")

        while queue:
            node = queue.popleft()
            print(node.data, end=" ")
            if node.left:
                queue.append(node.left)
            if node.right:
                queue.append(node.right)

        print()

# Example usage
if __name__ == "__main__":
    tree = Tree()
    s = input("Enter tree (level order, space separated): ")  
    # e.g. "2 3 4 5 6 7 N N 8 N N 9"
    tree.build_tree(s)
    tree.bfs()

File: Binary Tree/BFS.cpp
#include <bits/stdc++.h>
using namespace std;

// Node structure
struct Node {
    int data;
    Node* left;
    Node* right;
    Node(int val) : data(val), left(nullptr), right(nullptr) {}
};

class Tree {
private:
    Node* root;

public:
    Tree() : root(nullptr) {}

    // Build tree from space-separated level-order string ("N" = null)
    void buildTree(const string& s) {
        if (s.empty() || s[0] == 'N') return;

        vector<string> vals;
        stringstream ss(s);
        string word;
        while (ss >> word) vals.push_back(word);

        root = new Node(stoi(vals[0]));
        queue<Node*> q;
        q.push(root);

        int i = 1;
        while (!q.empty() && i < (int)vals.size()) {
            Node* curr = q.front(); q.pop();

            // left child
            if (vals[i] != "N") {
                curr->left = new Node(stoi(vals[i]));
                q.push(curr->left);
            }
            i++;

            // right child
            if (i < (int)vals.size() && vals[i] != "N") {
                curr->right = new Node(stoi(vals[i]));
                q.push(curr->right);
            }
            i++;
        }
    }

    // BFS / Level-order traversal
    void bfs() const {
        if (!root) {
            cout << "Tree is empty.\n";
            return;
        }

        queue<Node*> q;
        q.push(root);
        cout << "Level-order Traversal: ";

        while (!q.empty()) {
            Node* curr = q.front(); q.pop();
            cout << curr->data << " ";
            if (curr->left)  q.push(curr->left);
            if (curr->right) q.push(curr->right);
        }
        cout << "\n";
    }
};

int main() {
    Tree tree;
    cout << "Enter tree (level order, space-separated; use 'N' for null):\n";
    string input;
    getline(cin, input);

    tree.buildTree(input);
    tree.bfs();

    return 0;
}
